src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -1050,5 +1051,64 @@ private static TextEdit BuildRevertedEdit(TextEdit provisionalTextEdit)
             var formattingOptions = _formattingOptionsProvider.GetOptions(document.Uri);
             return Task.FromResult(formattingOptions);
         }
+
+        public override async Task<WorkspaceEdit?> RenameAsync(DelegatedRenameParams request, CancellationToken cancellationToken)
+        {
+            var hostDocumentUri = request.HostDocument.Uri;
+            if (!_documentManager.TryGetDocument(hostDocumentUri, out var documentSnapshot))
+            {
+                return null;
+            }
+
+            string languageServerName;
+            Uri projectedUri;
+            VirtualDocumentSnapshot virtualDocumentSnapshot;
+            if (request.ProjectedKind == RazorLanguageKind.Html &&
+                documentSnapshot.TryGetVirtualDocument<HtmlVirtualDocumentSnapshot>(out var htmlVirtualDocument))
+            {
+                languageServerName = RazorLSPConstants.HtmlLanguageServerName;
+                projectedUri = htmlVirtualDocument.Uri;
+                virtualDocumentSnapshot = htmlVirtualDocument;
+            }
+            else if (request.ProjectedKind == RazorLanguageKind.CSharp &&
+                documentSnapshot.TryGetVirtualDocument<CSharpVirtualDocumentSnapshot>(out var csharpVirtualDocument))
+            {
+                languageServerName = RazorLSPConstants.RazorCSharpLanguageServerName;
+                projectedUri = csharpVirtualDocument.Uri;
+                virtualDocumentSnapshot = csharpVirtualDocument;
+            }
+            else
+            {
+                Debug.Fail("Unexpected RazorLanguageKind. This shouldn't really happen in a real scenario.");
+                return null;
+            }
+
+            var synchronized = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync(
+                request.HostDocument.Version, virtualDocumentSnapshot, rejectOnNewerParallelRequest: false, cancellationToken);
+
+            if (!synchronized)
+            {
+                return null;
+            }
+
+            var renameParams = new RenameParams()
```

Comment:
This could be difficult to normalize but mayyyybeeee that could be encapsulated in some way